controllers/application_poller_controller.go

```diff
@@ -0,0 +1,288 @@
+			// TODO we should figure out the why for this
+			if activity.ID == "" {
+				continue
+			}
+
+			p.handleActivity(ctx, activity)
+		}
+	}
+}
+
+// handleActivity performs the task required for each activity.
+// When an ActivityItem is tagged with scan, the generation workflow is used to generate an experiment and the result
+// is converted into an api.Template consisting of parameters and metrics.
+// When an ActivityItem is tagged with run, the previous scanned template results are merged with
+// the results of an experiment generation workflow. Following this, the generated resources are applied/created
+// in the cluster.
+// note, rbac defined in cli/internal/commands/grant_permissions/generator
+func (p *Poller) handleActivity(ctx context.Context, activity applications.ActivityItem) {
+	// We always want to delete the activity after having received it
+	defer func() {
+		if err := p.apiClient.DeleteActivity(ctx, activity.URL); err != nil {
+			p.handleErrors(ctx, fmt.Errorf("%s: %w", "failed to delete activity", err), activity.URL)
+		}
+	}()
+
+	// Ensure we actually have an action to perform
+	if len(activity.Tags) != 1 {
+		p.handleErrors(ctx, fmt.Errorf("%s %d", "invalid number of activity tags, expected 1 got", len(activity.Tags)), activity.URL)
+		return
+	}
+
+	p.log.Info("starting activity task", "task", activity.Tags[0], "activity", activity.URL)
+
+	// Activity feed provides us with a scenario URL
+	scenario, err := p.apiClient.GetScenario(ctx, activity.ExternalURL)
+	if err != nil {
+		p.handleErrors(ctx, fmt.Errorf("%s (%s): %w", "failed to get scenario", activity.ExternalURL, err), activity.URL)
+		return
+	}
+
+	// Need to fetch top level application so we can get the resources
+	applicationURL := scenario.Link(api.RelationUp)
+	if applicationURL == "" {
+		p.handleErrors(ctx, fmt.Errorf("no matching application URL for scenario"), activity.URL)
+		return
+	}
+
+	templateURL := scenario.Link(api.RelationTemplate)
+	if templateURL == "" {
+		p.handleErrors(ctx, fmt.Errorf("no matching template URL for scenario"), activity.URL)
+		return
+	}
+
+	apiApp, err := p.apiClient.GetApplication(ctx, applicationURL)
+	if err != nil {
+		p.handleErrors(ctx, fmt.Errorf("%s (%s): %w", "failed to get application", activity.URL, err), activity.URL)
+		return
+	}
+
+	var assembledApp *optimizeappsv1alpha1.Application
+	if assembledApp, err = server.APIApplicationToClusterApplication(apiApp, scenario); err != nil {
+		p.handleErrors(ctx, fmt.Errorf("%s (%s): %w", "failed to assemble application", activity.URL, err), activity.URL)
+		return
+	}
+
+	generatedResources, err := p.generateApp(*assembledApp)
+	if err != nil {
+		p.handleErrors(ctx, fmt.Errorf("%s (%s): %w", "failed to generate application", activity.URL, err), activity.URL)
+		return
+	}
+
+	var exp *optimizev1beta2.Experiment
+	for i := range generatedResources {
+		if expObj, ok := generatedResources[i].(*optimizev1beta2.Experiment); ok {
+			exp = expObj
```

Comment:
Add annotations to experiment
- experiment url ( from experiments url link + generated name of experiment )